src/selectors/entities/teams.js

```diff
@@ -137,14 +138,75 @@ export function getTeamMember(state, teamId, userId) {
+    (state) => haveISystemPermission(state, {permission: Permissions.LIST_PRIVATE_TEAMS}),
+    (state) => state.entities.general.serverVersion,
+    (teams, myMembers, canListPublicTeams, canListPrivateTeams, serverVersion) => {
+        return Object.keys(teams).filter((id) => {
+            const team = teams[id];
+            const member = myMembers[id];
+            let canList = team.allow_open_invite;
+            if (isMinimumServerVersion(serverVersion, 5, 10, 0) ||
+                   (serverVersion.indexOf('dev') !== -1 && isMinimumServerVersion(serverVersion, 5, 8, 0)) ||
+                   (serverVersion.match(/^5.8.\d.\d\d\d\d.*$/) !== null && isMinimumServerVersion(serverVersion, 5, 8, 0))) {
+                canList = (canListPrivateTeams && !team.allow_open_invite) || (canListPublicTeams && team.allow_open_invite);
+            }
+            return team.delete_at === 0 && canList && !member;
+        });
+    }
+);
+
+export const getListableTeams = createSelector(
+    getTeams,
+    getListableTeamIds,
+    (teams, listableTeamIds) => {
+        return listableTeamIds.map((id) => teams[id]);
+    }
+);
+
+export const getSortedListableTeams = createSelector(
+    getTeams,
+    getListableTeamIds,
+    (state, locale) => locale,
+    (teams, listableTeamIds, locale) => {
+        const listableTeams = {};
+
+        for (const id of listableTeamIds) {
+            listableTeams[id] = teams[id];
+        }
+
+        function sortTeams(a, b) {
+            if (a.display_name !== b.display_name) {
+                return a.display_name.toLowerCase().localeCompare(b.display_name.toLowerCase(), locale || General.DEFAULT_LOCALE, {numeric: true});
+            }
+
+            return a.name.toLowerCase().localeCompare(b.name.toLowerCase(), locale || General.DEFAULT_LOCALE, {numeric: true});
+        }
+
+        return Object.values(listableTeams).sort(sortTeams);
+    }
+);
+
 export const getJoinableTeamIds = createIdsSelector(
     getTeams,
     getTeamMemberships,
-    (teams, myMembers) => {
+    (state) => haveISystemPermission(state, {permission: Permissions.JOIN_PUBLIC_TEAMS}),
+    (state) => haveISystemPermission(state, {permission: Permissions.JOIN_PRIVATE_TEAMS}),
+    (state) => state.entities.general.serverVersion,
+    (teams, myMembers, canJoinPublicTeams, canJoinPrivateTeams, serverVersion) => {
         return Object.keys(teams).filter((id) => {
             const team = teams[id];
             const member = myMembers[id];
-            return team.delete_at === 0 && team.allow_open_invite && !member;
+            let canJoin = team.allow_open_invite;
+            if (isMinimumServerVersion(serverVersion, 5, 10, 0) ||
+                   (serverVersion.indexOf('dev') !== -1 && isMinimumServerVersion(serverVersion, 5, 8, 0)) ||
+                   (serverVersion.match(/^5.8.\d.\d\d\d\d.*$/) !== null && isMinimumServerVersion(serverVersion, 5, 8, 0))) {
```

Comment:
Could this complex check be in a utility function? e.g.

```
if (isCompatibleWithJoinViewChannelPermissions(serverVersion)) {
```

Maybe that could itself call `serverVersionCheck(minReleaseVersion, minDevVersion)` so that all checks of this sort can use that function to avoid copy/pasting all the matching magic for dev/rc builds?